aip/0233.md

```diff
@@ -0,0 +1,82 @@
+---
+aip:
+  id: 233
+  state: draft
+  created: 2019-06-18
+permalink: /233
+redirect_from:
+  - /0233
+---
+
+# Batch Create
+
+Some APIs need to allow users to create multiple resources in a single
+transaction. A batch create method provides this functionality.
+
+## Guidance
+
+Batch create methods are specified using the following pattern:
+
+```proto
+rpc BatchCreateBooks(BatchCreateBooksRequest) returns (BatchCreateBooksResponse) {
+  option (google.api.http) = {
+    post: "/v1/shelves/-/books:batchCreate"
+    body: "*"
+  };
+}
+```
+
+- The RPC's name **must** begin with `BatchCreate`. The remainder of
+  the RPC name **should** be the plural form of the resource being created.
+- The request and response messages **must** match the RPC name, with `-Request`
+  and `-Response` suffixes.
+  - However, in the event that the request may take a significant amount of
+    time, the response message **must** be a `google.longrunning.Operation`
+    which ultimately resolves to the `-Response` type.
+- The HTTP verb **must** be `POST`.
+- The HTTP URI **must** end with `:batchCreate`.
+- The URL path **should** represent the collection for the resource, matching
+  the collection used for simple CRUD operations. If the operation spans
+  parents, a dash (`-`) **may** be used as a wildcard.
+- The body clause in the `google.api.http` annotation **should** be `"*"`.
+- The operation **should** be atomic: it **should** fail for all resources or
+  succeed for all resources (no partial success).
+
+### Request message
+
+The request for a batch create method **should** be specified with the following pattern:
+
+```proto
+message BatchCreateBookRequest {
+  // The request message specifying the resources to create.
+  // A maximum of 1000 books can be created in a batch.
+  repeated CreateBookRequest requests = 1;
```

Comment:
Do we want to allow (encourage? require?) `repeated Book books` here?
This would be easier for services that do not need customer-provided IDs.

Reply:
True, but it pushes some types of functionality additions into breaking changes instead of additive changes (basically anything that would be specified in a Create request)

Reply:
@jgeewax and I discussed this and we landed on the position taken in the existing text.